filebeat/input/file/state.go

```diff
@@ -81,25 +83,36 @@ func (s *States) findPrevious(newState State) (int, State) {
 }
 
 // Cleanup cleans up the state array. All states which are older then `older` are removed
-func (s *States) Cleanup(older time.Duration) {
+func (s *States) Cleanup() {
+
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	for i, state := range s.states {
 
-		// File wasn't seen for longer then older -> remove state
-		if time.Since(state.LastSeen) > older {
-			logp.Debug("prospector", "State removed for %s because of older: %s", state.Source)
-			s.states = append(s.states[:i], s.states[i+1:]...)
+		// -1 means ttl is disabled
+		if state.Ttl == -1 {
+			continue
 		}
-	}
 
+		// States is older then ttl
+		if int64(time.Since(state.Timestamp).Seconds()) > state.Ttl {
+			// This check prevents a panic
+			if len(s.states) == 1 {
+				s.states = []State{}
+			} else {
+				s.states = append(s.states[:i], s.states[i+1:]...)
+			}
```

Comment:
maybe get timestamp before removing, so all states are compared against same timestamp.

Are removed states somehow reported to registrar for remove? If so, how can we be sure some state is not updated by some prospector?